### PR TITLE
Fence PCI atomics

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by shamisp** [comment]
I would like to get some description for the patch. What PCIe problem are you trying to solve here ?

**Comment by yosefe** [comment]
PCIe AtomicOps are unordered. This patch forces ordering per destination, when fence is used.

**Comment by shamisp** [comment]
How underlaying fence is actually implemented ? Drain CQ ?

**Comment by yosefe** [comment]
It waits for previous WQEs to be completed (AFAIR it's defined in IB spec)

**Comment by shamisp** [comment]
Do you set fence bit in WQE ?

**Comment by yosefe** [comment]
yes

**Comment by shamisp** [comment]
This is not part of this code ? can you please point out to the code that set the bit ?

**Comment by yosefe** [comment]
it's set in a device-specific way

**Comment by shamisp** [comment]
Hmmm, AFAIK there is WQE field to set this ?

**Comment by yosefe** [comment]
This PR is for mlx5 transport only

**Comment by shamisp** [comment]
but the same has to be done for regular verbs, right ?

**Comment by yosefe** [comment]
yes, in another PR

**Comment by shamisp** [comment]
Ok. So , you don't have to wait for WQE to be completed. You just post it with fence flag on WQE and this is it.

I guess, this has to be somewhat tunable. Let's assume there is IB/ROCE device that is integrated on SoC and does not have to go through PCIe ? In this case, I guess you don't need the fence ?

**Comment by yosefe** [comment]
yes, on Verbs transport it will be configurable option. on mlx5 we can auto-detect it

**Comment by shamisp** [comment]
👍

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8796/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8840/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8846/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
RFC @yosefe

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8901/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8907/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8908/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8909/ for details (Mellanox internal link).

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 1]
```diff
@@ -26,17 +26,22 @@ uct_dc_mlx5_iface_bcopy_post(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep,
                             uint32_t imm_val_be, const void *buffer,
                             uct_ib_log_sge_t *log_sge)
 {
+    uint8_t fm_ce_se;
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
     desc->super.sn = txwq->sw_pi;
+    fm_ce_se = send_flags;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: seems this logic should be moved to uct_rc_mlx5_txqp_dptr_post[_iov] ?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 2]
```diff
@@ -26,17 +26,22 @@ uct_dc_mlx5_iface_bcopy_post(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep,
...
+    if (opcode == MLX5_OPCODE_RDMA_READ) {
+        fm_ce_se |= uct_rc_mlx5_ep_atomic_fence(&iface->super, txwq);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: rename uct_rc_mlx5_ep_atomic_fence() to uct_rc_mlx5_ep_fm()

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 3]
```diff
@@ -222,6 +233,21 @@ ucs_status_t uct_dc_mlx5_ep_atomic32_fetch(uct_ep_h ep, uct_atomic_op_t opcode,
                                           remote_addr, rkey, comp);
 }
 
+ucs_status_t uct_dc_mlx5_ep_fence(uct_ep_h tl_ep, unsigned flags)
+{
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
+    uct_dc_mlx5_ep_t *ep = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uct_ib_md_t *md = uct_ib_iface_md(&iface->super.super.super);
+
+    if ((UCT_IB_MLX5_MD_FLAGS(md) & UCT_IB_MLX5_MD_FLAG_PCI_ATOMIC) &&
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: pls add comment that need fence for pci atomic

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 4]
```diff
@@ -206,6 +213,8 @@ typedef struct uct_ib_mlx5_txwq {
 #if ENABLE_ASSERT
     uint16_t                    hw_ci;
 #endif
+    uint16_t                    fence_beat;
+    uint8_t                     next_fence;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: next_fm (because fm is the name in PRM)

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 5]
```diff
@@ -1296,3 +1296,19 @@ uct_rc_mlx5_iface_common_atomic_data(unsigned opcode, unsigned size, uint64_t va
...
 }
+
+static UCS_F_ALWAYS_INLINE uint8_t
+uct_rc_mlx5_ep_atomic_fence(uct_rc_mlx5_iface_common_t *iface,
+                            uct_ib_mlx5_txwq_t *txwq)
+{
+    uint8_t fm_ce_se = MLX5_WQE_CTRL_CQ_UPDATE;
+
+    fm_ce_se |= txwq->next_fence;
+    if (txwq->fence_beat != iface->tx.fence_beat) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl]: pls add comment to explain the algorithm

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8916/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8918/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8929/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8930/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8932/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/rc/base/rc_iface.c [hunk 6]
```diff
@@ -178,6 +178,13 @@ ucs_status_t uct_rc_iface_query(uct_rc_iface_t *iface,
     }
 #endif
 
+    if (dev->flags & UCT_IB_DEVICE_FLAG_PCI_ATOMICS) {
+        iface_attr->cap.atomic32.op_flags = 0;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: align on column

File: src/uct/ib/rc/base/rc_iface.c [hunk 7]
```diff
@@ -178,6 +178,13 @@ ucs_status_t uct_rc_iface_query(uct_rc_iface_t *iface,
...
+        iface_attr->cap.atomic32.fop_flags = 0;
+        iface_attr->cap.atomic64.op_flags = 0;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: why are all atomics disabled now? is it going to be fixed in next PR?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 8]
```diff
@@ -409,6 +409,8 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_common_t,
     self->tx.bb_max                  = ucs_min(config->tx_max_bb, UINT16_MAX);
     self->super.config.tx_moderation = ucs_min(self->super.config.tx_moderation,
                                                self->tx.bb_max / 4);
+    self->tx.next_fm = 0;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: minor: align = on column

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 9]
```diff
@@ -214,6 +214,23 @@ uct_rc_mlx5_iface_common_am_handler(uct_rc_mlx5_iface_common_t *iface,
...
 }
 
+static UCS_F_ALWAYS_INLINE uint8_t
+uct_rc_mlx5_ep_fm(uct_rc_mlx5_iface_common_t *iface, uct_ib_mlx5_txwq_t *txwq)
+{
+    uint8_t fm_ce_se = MLX5_WQE_CTRL_CQ_UPDATE;
+
+    fm_ce_se |= txwq->next_fm;
+    txwq->next_fm = 0;
+    /* iface increase beat each time iface_fence requested, so, if our beat
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl]: space line before

Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5.inl]: /* a call to iface_fence increases beat, so if endpoint beat is not in sync with iface beat it means the endpoint did not post any WQE with fence flag yet */

File: src/uct/ib/rc/base/rc_iface.c (hunk 7 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: yes

File: src/uct/ib/base/ib_md.c [hunk 10]
```diff
@@ -1676,6 +1676,12 @@ static ucs_status_t uct_ib_verbs_md_open(struct ibv_device *ibv_device,
     }
 #endif
 
+#if HAVE_DECL_IBV_EXP_DEVICE_ATTR_PCI_ATOMIC_CAPS
+    if (dev->dev_attr.pci_atomic_caps.fetch_add) {
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c]: currently both are 2 if pci atmic enabled and 0 if disabled

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 11]
```diff
@@ -298,6 +298,10 @@ static ucs_status_t uct_ib_mlx5dv_md_open(struct ibv_device *ibv_device,
             }
             dev->ext_atomic_arg_sizes_be = arg_size;
         }
+
+        if (UCT_IB_MLX5DV_GET(atomic_caps, cap, fetch_add_pci_atomic)) {
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5_dv.c]: currently both are 2 if pci atmic enabled and 0 if disabled

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 12]
```diff
@@ -730,6 +731,15 @@ void uct_rc_mlx5_iface_common_query(uct_ib_iface_t *ib_iface,
         iface_attr->cap.flags              |= UCT_IFACE_FLAG_ATOMIC_DEVICE;
     }
 
+    if (md->dev.flags & UCT_IB_DEVICE_FLAG_PCI_ATOMICS) {
+        iface_attr->cap.atomic64.op_flags  &= UCT_RC_MLX5_PCI_ATOMIC_OPS;
+        iface_attr->cap.atomic64.fop_flags &= UCT_RC_MLX5_PCI_ATOMIC_OPS;
+        iface_attr->cap.atomic32.op_flags  &= UCT_RC_MLX5_PCI_ATOMIC_OPS;
+        iface_attr->cap.atomic32.fop_flags &= UCT_RC_MLX5_PCI_ATOMIC_OPS;
+        iface_attr->cap.flags              &= ~UCT_IFACE_FLAG_ATOMIC_DEVICE;
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: there are multiple places which set, check PCI_ATOMICS each time?

File: src/uct/ib/base/ib_md.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: so let's have || condition here

File: src/uct/ib/mlx5/ib_mlx5_dv.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5_dv.c]: add ||

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5_dv.c]: we don't want to rely on particular implementation, but be more compliant with PRM/API definitions (in this case, it's not so hard to have it)

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: i see only uct_rc_iface_query and uct_rc_mlx5_iface_common_query, anything else?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8963/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 13]
```diff
@@ -1677,8 +1677,11 @@ static ucs_status_t uct_ib_verbs_md_open(struct ibv_device *ibv_device,
 #endif
 
 #if HAVE_DECL_IBV_EXP_DEVICE_ATTR_PCI_ATOMIC_CAPS
-    if (dev->dev_attr.pci_atomic_caps.fetch_add) {
+    if (dev->dev_attr.pci_atomic_caps.fetch_add ||
+        dev->dev_attr.pci_atomic_caps.compare_swap) {
         dev->flags |= UCT_IB_DEVICE_FLAG_PCI_ATOMICS;
+        dev->pci_fadd_arg_sizes = dev->dev_attr.pci_atomic_caps.fetch_add << 2;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: minor: align = on column

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 14]
```diff
@@ -299,8 +299,11 @@ static ucs_status_t uct_ib_mlx5dv_md_open(struct ibv_device *ibv_device,
             dev->ext_atomic_arg_sizes_be = arg_size;
         }
 
-        if (UCT_IB_MLX5DV_GET(atomic_caps, cap, fetch_add_pci_atomic)) {
+        if (UCT_IB_MLX5DV_GET(atomic_caps, cap, fetch_add_pci_atomic) ||
+            UCT_IB_MLX5DV_GET(atomic_caps, cap, compare_swap_pci_atomic)) {
             dev->flags |= UCT_IB_DEVICE_FLAG_PCI_ATOMICS;
+            dev->pci_fadd_arg_sizes = UCT_IB_MLX5DV_GET(atomic_caps, cap, fetch_add_pci_atomic) << 2;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: can we remove UCT_IB_DEVICE_FLAG_PCI_ATOMICS now? and rely on (dev->pci_fadd_arg_sizes || dev->pci_cswap_arg_sizes)?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8964/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8970/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8981/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8985/ for details (Mellanox internal link).

**Pull request merged**